lib/Dialect/LLHD/Transforms/TemporalCodeMotionPass.cpp

```diff
@@ -88,6 +88,45 @@ getBranchDecisionsFromDominatorToTarget(OpBuilder &builder, Block *driveBlock,
   return mem[driveBlock];
 }
 
+/// Move 'llhd.drv' op '$value' definition (if it present) and it's dependencies
+/// by lookup in current block. It's expected that other dependencies are placed
+/// before 'moveBefore'. It's almost 'true' due to 'hw.constant' and 'llhd.sig'
+/// ops are placed before 'llhd.process'.
+static void moveValueDependencies(llhd::DrvOp drvOp, Operation *moveBefore) {
+  auto *drvBlock = drvOp->getBlock();
+  auto value = drvOp.getValue();
+  if (value.getParentBlock() != drvBlock || !value.getDefiningOp())
+    return;
+
+  llvm::SmallVector<Operation *> moveSet;
+  moveSet.push_back(value.getDefiningOp());
+
+  // Algorithm is finite due to recursive dependencies can't exist
+  // in single linear block
+  for (unsigned currOp = 0; currOp != moveSet.size(); currOp++) {
+    for (auto arg : moveSet[currOp]->getOperands()) {
+      if (arg.getParentBlock() == drvBlock) {
+        auto *op = arg.getDefiningOp();
+        assert(op && "Using block argument as llhd.drv value is unexpected");
```

Comment:
This assumes that passes like `block-argument-to-mux` are called beforehand. But the pass shouldn't crash if that's not the case, i.e., this should fail with a proper error message or be supported.

Reply:
> I was thinking about merging
Can I try to implement this? Should old passes be deprecated or removed? Should corresponding files be merged or can we just call passes successively in third file?